### ISSUE TITLE
Mismatch in current and updated revision

### ISSUE THREAD

**Issue by saketmht** [Opster/opensearch-k8s-operator#605]
When first creating the cluster or on any change to the CR, it seems that sometimes the sts is not able to be fully ready. By ready I mean that readyReplicas == updatedReplicas == spec.Replicas and currentRevision == updatedRevision.

After applying the below CR, while the pods are all up and running, the master node sts has this status
```
status:
  availableReplicas: 3
  collisionCount: 0
  currentReplicas: 1
  currentRevision: my-first-cluster-masters-76c597cf48
  observedGeneration: 2
  readyReplicas: 3
  replicas: 3
  updateRevision: my-first-cluster-masters-76954f844b
  updatedReplicas: 2
```

Data sts and ingest sts are full ready. But sometimes, other sts are not ready while master is ready
```
status:
  availableReplicas: 3
  collisionCount: 0
  currentReplicas: 3
  currentRevision: my-first-cluster-data-6584c66d69
  observedGeneration: 1
  readyReplicas: 3
  replicas: 3
  updateRevision: my-first-cluster-data-6584c66d69
  updatedReplicas: 3
```

Ideally, the operator should see that all the pods are not up to date and restart the outdated pod to match the update revision but I see no such action happening. 
Other details:

1. It doesn't seem that the operator is even trying to do a rolling restart or any restart of any kind.
2. The cluster health is green. No shards are unallocated.
3. All the pods are up and running.

```
apiVersion: opensearch.opster.io/v1
kind: OpenSearchCluster
metadata:
  name: my-first-cluster
  namespace: default
spec:
  security:
    config:
    tls:
       http:
         generate: true
       transport:
         generate: true
  general:
    httpPort: 9200
    serviceName: my-first-cluster
    version: 2.3.0
    drainDataNodes: true
  dashboards:
    tls:
      enable: true
      generate: true
    version: 2.3.0
    enable: true
    replicas: 1
    resources:
      requests:
         memory: "512Mi"
         cpu: "200m"
      limits:
         memory: "512Mi"
         cpu: "200m"
  nodePools:
    - component: masters
      replicas: 3
      diskSize: "50Gi"
      resources:
         requests:
            memory: "1.4Gi"
      roles:
        - "cluster_manager"
    - component: data
      replicas: 3
      diskSize: "50Gi"
      resources:
         requests:
            memory: "4.8Gi"
      roles:
        - "data"
    - component: ingest
      replicas: 1
      resources:
         requests:
            memory: "2.5Gi"
      roles:
        - "ingest"
      persistence:
        emptyDir: { }
```

**Comment by saketmht** [comment]
Not sure if these relate to :
https://github.com/Opster/opensearch-k8s-operator/issues/604
https://github.com/Opster/opensearch-k8s-operator/issues/310

**Comment by saketmht** [comment]
On further investigation, this is the case for nodes other than data nodes because of this check in rollingRestart.go

```
for _, nodePool := range r.instance.Spec.NodePools {
		if helpers.HasDataRole(&nodePool) {
			sts := &appsv1.StatefulSet{}
			if err := r.Get(r.ctx, types.NamespacedName{
				Name:      builders.StsName(r.instance, &nodePool),
				Namespace: r.instance.Namespace,
			}, sts); err != nil {
				return ctrl.Result{}, err
			}
			if sts.Status.ReadyReplicas != pointer.Int32Deref(sts.Spec.Replicas, 1) {
				return ctrl.Result{
					Requeue:      true,
					RequeueAfter: 10 * time.Second,
				}, nil
			}

			if sts.Status.UpdateRevision != "" &&
				sts.Status.UpdatedReplicas != pointer.Int32Deref(sts.Spec.Replicas, 1) {
				pendingUpdate = true
			}
		}
	}
```

We only seem to rolling restart for data node sts?? How are other updates to other sts handled?

**Comment by swoehrl-mw** [comment]
Should be fixed with PR #614. Fix will be part of next release.
